curvefs/src/mds/heartbeat/copyset_conf_generator.cpp

```diff
@@ -0,0 +1,193 @@
+                LOG(WARNING) << "topoUpdater update copyset("
+                             << reportCopySetInfo.GetPoolId() << ","
+                             << reportCopySetInfo.GetId()
+                             << ") got error code: " << updateCode;
+                return false;
+            } else {
+                // update to memory successfully
+                return true;
+            }
+        } else {
+            // no config to dispatch
+            return false;
+        }
+    } else {
+        return FollowerGenCopysetConf(reportId, reportCopySetInfo,
+                                      recordCopySetInfo, copysetConf);
+    }
+}
+
+MetaServerIdType CopysetConfGenerator::LeaderGenCopysetConf(
+    const ::curvefs::mds::topology::CopySetInfo &copySetInfo,
+    const ::curvefs::mds::heartbeat::ConfigChangeInfo &configChInfo,
+    ::curvefs::mds::heartbeat::CopySetConf *copysetConf) {
+    // pass date to scheduler
+    return coordinator_->CopySetHeartbeat(copySetInfo, configChInfo,
+                                          copysetConf);
+}
+
+bool CopysetConfGenerator::FollowerGenCopysetConf(
+    MetaServerIdType reportId,
+    const ::curvefs::mds::topology::CopySetInfo &reportCopySetInfo,
+    const ::curvefs::mds::topology::CopySetInfo &recordCopySetInfo,
+    ::curvefs::mds::heartbeat::CopySetConf *copysetConf) {
+    // if there's no candidate on a copyset, and epoch on MDS is larger or equal
+    // to what non-leader copy report,                //NOLINT
+    // copy(ies) can be deleted according to the configuration of MDS
+
+    // epoch that MDS recorded >= epoch reported
+    if (recordCopySetInfo.GetEpoch() >= reportCopySetInfo.GetEpoch()) {
+        // delay cleaning is for avoiding cases like the example below:
+        // the initial config of a copyset is ABC, in which A is the leader, and
+        // epoch = 8
+        // 1. MDS gereration an operator ABC+D and dispatch to A
+        // 2. A installed snapshot to D successfully, and start copying journal
+        // 3. When D is copying the journal, old config of the copyset is
+        // applied (e.g. ABE, epoch = 5)
+        // 4. MDS restart now and operator is lost
+        // 5. After the MDS has restarted, it first receive heartbeat from
+        // follower-id (ABE, epoch = 5).
+        //    In normal cases (without delay cleaning), this copyset on D will
+        //    be deleted since its epoch number is lower,
+        //    but it will be a problem in our case since D is already a member
+        //    of the copyset, and the deletion of D means
+        //    the elimination of copies.
+        // here's how delay cleaning works:
+        //    In one way leader will report the existence of D, which promise
+        //    that D will not be deleted by mistake. In another
+        //    way, the configuration that D report will be up-to-date since the
+        //    journal replay has completed
```

Comment:
this annotation is the same as header file.

Reply:
done